File: apps/android/app/src/main/java/com/signalasi/chat/SignalASILinkTransportDiagnostics.kt
package com.signalasi.chat

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest
import java.util.Locale
import java.util.UUID

internal enum class SignalASILinkDiagnosticKind(val wireName: String) {
    ENCRYPTED_REPLAY("encrypted_replay"),
    PENDING_REPLAY("pending_replay"),
    DUPLICATE_MESSAGE("duplicate_message"),
    DUPLICATE_RECEIPT("duplicate_receipt"),
    OLD_COUNTER("old_counter"),
    DECRYPT_FAILURE("decrypt_failure"),
    CHUNK_DUPLICATE("chunk_duplicate"),
    FRAGMENT_REJECTED("fragment_rejected");

    companion object {
        fun fromWireName(value: String): SignalASILinkDiagnosticKind? =
            entries.firstOrNull { it.wireName == value }
    }
}

internal data class SignalASILinkDiagnosticEvent(
    val id: String,
    val kind: SignalASILinkDiagnosticKind,
    val recordedAtMillis: Long,
    val endpointRef: String,
    val messageRef: String,
    val detailCode: String
)

internal data class SignalASILinkDiagnosticSnapshot(
    val totalEvents: Long,
    val counts: Map<SignalASILinkDiagnosticKind, Long>,
    val recentEvents: List<SignalASILinkDiagnosticEvent>
) {
    val replayCount: Long
        get() = count(SignalASILinkDiagnosticKind.ENCRYPTED_REPLAY) +
            count(SignalASILinkDiagnosticKind.PENDING_REPLAY)

    val duplicateCount: Long
        get() = count(SignalASILinkDiagnosticKind.DUPLICATE_MESSAGE) +
            count(SignalASILinkDiagnosticKind.DUPLICATE_RECEIPT) +
            count(SignalASILinkDiagnosticKind.CHUNK_DUPLICATE)

    val oldCounterCount: Long
        get() = count(SignalASILinkDiagnosticKind.OLD_COUNTER)

    val failureCount: Long
        get() = count(SignalASILinkDiagnosticKind.DECRYPT_FAILURE) +
            count(SignalASILinkDiagnosticKind.FRAGMENT_REJECTED)

    fun count(kind: SignalASILinkDiagnosticKind): Long = counts[kind] ?: 0L
}

internal interface SignalASILinkDiagnosticStore {
    fun read(): JSONObject
    fun write(value: JSONObject)
    fun clear()
}

internal class SharedPreferencesSignalASILinkDiagnosticStore(
    context: Context
) : SignalASILinkDiagnosticStore {
    private val preferences: SharedPreferences = context.applicationContext.getSharedPreferences(
        PREFERENCES,
        Context.MODE_PRIVATE
    )

    override fun read(): JSONObject = runCatching {
        JSONObject(preferences.getString(KEY_STATE, null) ?: "{}")
    }.getOrDefault(JSONObject())

    override fun write(value: JSONObject) {
        preferences.edit().putString(KEY_STATE, value.toString()).apply()
    }

    override fun clear() {
        preferences.edit().clear().apply()
    }

    companion object {
        const val PREFERENCES = "signalasi_link_transport_diagnostics"
        private const val KEY_STATE = "state"
    }
}

internal class InMemorySignalASILinkDiagnosticStore : SignalASILinkDiagnosticStore {
    private var state = JSONObject()

    override fun read(): JSONObject = JSONObject(state.toString())

    override fun write(value: JSONObject) {
        state = JSONObject(value.toString())
    }

    override fun clear() {
        state = JSONObject()
    }
}

internal class SignalASILinkDiagnosticLedger(
    private val store: SignalASILinkDiagnosticStore,
    private val clock: () -> Long = System::currentTimeMillis,
    private val maximumEvents: Int = DEFAULT_MAXIMUM_EVENTS
) {
    init {
        require(maximumEvents > 0)
    }

    @Synchronized
    fun record(
        kind: SignalASILinkDiagnosticKind,
        endpointIdentity: String = "",
        messageIdentity: String = "",
        detailCode: String = ""
    ): SignalASILinkDiagnosticSnapshot {
        val state = store.read()
        val counts = state.optJSONObject(KEY_COUNTS) ?: JSONObject()
        counts.put(kind.wireName, counts.optLong(kind.wireName, 0L) + 1L)

        val events = state.optJSONArray(KEY_EVENTS) ?: JSONArray()
        events.put(
            JSONObject()
                .put("id", UUID.randomUUID().toString())
                .put("kind", kind.wireName)
                .put("recorded_at", clock())
                .put("endpoint_ref", anonymizedReference(endpointIdentity))
                .put("message_ref", anonymizedReference(messageIdentity))
                .put("detail_code", normalizedDetailCode(detailCode))
        )
        while (events.length() > maximumEvents) {
            events.remove(0)
        }

        state.put(KEY_PROTOCOL, PROTOCOL)
        state.put(KEY_TOTAL, state.optLong(KEY_TOTAL, 0L) + 1L)
        state.put(KEY_COUNTS, counts)
        state.put(KEY_EVENTS, events)
        store.write(state)
        return snapshotOf(state)
    }

    @Synchronized
    fun snapshot(): SignalASILinkDiagnosticSnapshot = snapshotOf(store.read())

    @Synchronized
    fun clear() {
        store.clear()
    }

    private fun snapshotOf(state: JSONObject): SignalASILinkDiagnosticSnapshot {
        val countsObject = state.optJSONObject(KEY_COUNTS) ?: JSONObject()
        val counts = SignalASILinkDiagnosticKind.entries.associateWith { kind ->
            countsObject.optLong(kind.wireName, 0L)
        }
        val eventArray = state.optJSONArray(KEY_EVENTS) ?: JSONArray()
        val events = buildList {
            for (index in eventArray.length() - 1 downTo 0) {
                val item = eventArray.optJSONObject(index) ?: continue
                val kind = SignalASILinkDiagnosticKind.fromWireName(item.optString("kind")) ?: continue
                add(
                    SignalASILinkDiagnosticEvent(
                        id = item.optString("id"),
                        kind = kind,
                        recordedAtMillis = item.optLong("recorded_at"),
                        endpointRef = item.optString("endpoint_ref"),
                        messageRef = item.optString("message_ref"),
                        detailCode = item.optString("detail_code")
                    )
                )
            }
        }
        return SignalASILinkDiagnosticSnapshot(
            totalEvents = state.optLong(KEY_TOTAL, counts.values.sum()),
            counts = counts,
            recentEvents = events
        )
    }

    companion object {
        const val PROTOCOL = "signalasi.link-transport-diagnostics/1.0"
        const val DEFAULT_MAXIMUM_EVENTS = 40
        private const val KEY_PROTOCOL = "protocol"
        private const val KEY_TOTAL = "total_events"
        private const val KEY_COUNTS = "counts"
        private const val KEY_EVENTS = "recent_events"

        internal fun anonymizedReference(value: String): String {
            if (value.isBlank()) return ""
            return MessageDigest.getInstance("SHA-256")
                .digest(value.toByteArray(Charsets.UTF_8))
                .take(6)
                .joinToString("") { "%02x".format(Locale.US, it) }
        }

        internal fun normalizedDetailCode(value: String): String =
            value.trim()
                .lowercase(Locale.US)
                .replace(Regex("[^a-z0-9_.-]+"), "_")
                .trim('_')
                .take(64)
    }
}

internal object SignalASILinkTransportDiagnostics {
    @Volatile
    private var ledger: SignalASILinkDiagnosticLedger? = null

    fun record(
        context: Context,
        kind: SignalASILinkDiagnosticKind,
        endpointIdentity: String = "",
        messageIdentity: String = "",
        detailCode: String = ""
    ): SignalASILinkDiagnosticSnapshot =
        runtime(context).record(kind, endpointIdentity, messageIdentity, detailCode)

    fun snapshot(context: Context): SignalASILinkDiagnosticSnapshot =
        runtime(context).snapshot()

    fun clear(context: Context) {
        runtime(context).clear()
    }

    internal fun classifyDecryptionFailure(error: Throwable): SignalASILinkDiagnosticKind {
        val className = error.javaClass.simpleName.lowercase(Locale.US)
        val message = error.message.orEmpty().lowercase(Locale.US)
        return when {
            "old counter" in message || "oldcounter" in message ->
                SignalASILinkDiagnosticKind.OLD_COUNTER
            "duplicatemessage" in className || "duplicate message" in message ->
                SignalASILinkDiagnosticKind.DUPLICATE_MESSAGE
            else -> SignalASILinkDiagnosticKind.DECRYPT_FAILURE
        }
    }

    internal fun classifyFragmentFailure(error: Throwable): SignalASILinkDiagnosticKind =
        if ("duplicate" in error.message.orEmpty().lowercase(Locale.US)) {
            SignalASILinkDiagnosticKind.CHUNK_DUPLICATE
        } else {
            SignalASILinkDiagnosticKind.FRAGMENT_REJECTED
        }

    @Synchronized
    private fun runtime(context: Context): SignalASILinkDiagnosticLedger =
        ledger ?: SignalASILinkDiagnosticLedger(
            SharedPreferencesSignalASILinkDiagnosticStore(context.applicationContext)
        ).also { ledger = it }
}

File: apps/ios/SignalASI/SignalASILinkTransportDiagnostics.swift
import CryptoKit
import Foundation

enum SignalASILinkDiagnosticKind: String, Codable, CaseIterable {
  case encryptedReplay = "encrypted_replay"
  case pendingReplay = "pending_replay"
  case duplicateMessage = "duplicate_message"
  case duplicateReceipt = "duplicate_receipt"
  case oldCounter = "old_counter"
  case decryptFailure = "decrypt_failure"
  case chunkDuplicate = "chunk_duplicate"
  case fragmentRejected = "fragment_rejected"

  static func fromWireName(_ value: String) -> SignalASILinkDiagnosticKind? {
    allCases.first { $0.rawValue == value }
  }
}

struct SignalASILinkDiagnosticEvent: Codable, Equatable, Identifiable {
  var id: String
  var kind: SignalASILinkDiagnosticKind
  var recordedAtMillis: Int64
  var endpointRef: String
  var messageRef: String
  var detailCode: String

  enum CodingKeys: String, CodingKey {
    case id
    case kind
    case recordedAtMillis = "recorded_at"
    case endpointRef = "endpoint_ref"
    case messageRef = "message_ref"
    case detailCode = "detail_code"
  }
}

struct SignalASILinkDiagnosticSnapshot: Equatable {
  var totalEvents: Int64
  var counts: [SignalASILinkDiagnosticKind: Int64]
  var recentEvents: [SignalASILinkDiagnosticEvent]

  var replayCount: Int64 {
    count(.encryptedReplay) + count(.pendingReplay)
  }

  var duplicateCount: Int64 {
    count(.duplicateMessage) + count(.duplicateReceipt) + count(.chunkDuplicate)
  }

  var oldCounterCount: Int64 {
    count(.oldCounter)
  }

  var failureCount: Int64 {
    count(.decryptFailure) + count(.fragmentRejected)
  }

  func count(_ kind: SignalASILinkDiagnosticKind) -> Int64 {
    counts[kind] ?? 0
  }
}

protocol SignalASILinkDiagnosticStore {
  func read() -> Data?
  func write(_ data: Data)
  func clear()
}

final class UserDefaultsSignalASILinkDiagnosticStore: SignalASILinkDiagnosticStore {
  private let defaults: UserDefaults
  private let storageKey = "signalasi-ios-link-transport-diagnostics-v1"

  init(defaults: UserDefaults = .standard) {
    self.defaults = defaults
  }

  func read() -> Data? {
    defaults.data(forKey: storageKey)
  }

  func write(_ data: Data) {
    defaults.set(data, forKey: storageKey)
  }

  func clear() {
    defaults.removeObject(forKey: storageKey)
  }
}

final class InMemorySignalASILinkDiagnosticStore: SignalASILinkDiagnosticStore {
  private var data: Data?

  func read() -> Data? {
    data
  }

  func write(_ data: Data) {
    self.data = data
  }

  func clear() {
    data = nil
  }
}

final class SignalASILinkDiagnosticLedger {
  static let protocolVersion = "signalasi.link-transport-diagnostics/1.0"
  static let defaultMaximumEvents = 40

  private let store: SignalASILinkDiagnosticStore
  private let clockMillis: () -> Int64
  private let maximumEvents: Int
  private let lock = NSLock()

  init(
    store: SignalASILinkDiagnosticStore,
    clockMillis: @escaping () -> Int64 = {
      Int64((Date().timeIntervalSince1970 * 1_000).rounded())
    },
    maximumEvents: Int = defaultMaximumEvents
  ) {
    precondition(maximumEvents > 0)
    self.store = store
    self.clockMillis = clockMillis
    self.maximumEvents = maximumEvents
  }

  @discardableResult
  func record(
    kind: SignalASILinkDiagnosticKind,
    endpointIdentity: String = "",
    messageIdentity: String = "",
    detailCode: String = ""
  ) -> SignalASILinkDiagnosticSnapshot {
    lock.lock()
    defer { lock.unlock() }

    var state = readState()
    state.protocolName = Self.protocolVersion
    state.totalEvents += 1
    state.counts[kind.rawValue, default: 0] += 1
    state.recentEvents.append(SignalASILinkDiagnosticEvent(
      id: UUID().uuidString,
      kind: kind,
      recordedAtMillis: clockMillis(),
      endpointRef: Self.anonymizedReference(endpointIdentity),
      messageRef: Self.anonymizedReference(messageIdentity),
      detailCode: Self.normalizedDetailCode(detailCode)
    ))
    if state.recentEvents.count > maximumEvents {
      state.recentEvents = Array(state.recentEvents.suffix(maximumEvents))
    }
    writeState(state)
    return snapshot(of: state)
  }

  func snapshot() -> SignalASILinkDiagnosticSnapshot {
    lock.lock()
    defer { lock.unlock() }
    return snapshot(of: readState())
  }

  func clear() {
    lock.lock()
    defer { lock.unlock() }
    store.clear()
  }

  static func anonymizedReference(_ value: String) -> String {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
      return ""
    }
    let digest = Data(SHA256.hash(data: Data(trimmed.utf8)))
    return Data(digest.prefix(6)).hexString()
  }

  static func normalizedDetailCode(_ value: String) -> String {
    let lowered = value
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .lowercased(with: Locale(identifier: "en_US_POSIX"))
      .replacingOccurrences(of: "[^a-z0-9_.-]+", with: "_", options: .regularExpression)
      .trimmingCharacters(in: CharacterSet(charactersIn: "_"))
    return String(lowered.prefix(64))
  }

  private func readState() -> PersistedState {
    guard let data = store.read(),
          let state = try? JSONDecoder.linkDiagnostics.decode(PersistedState.self, from: data) else {
      return PersistedState()
    }
    return state
  }

  private func writeState(_ state: PersistedState) {
    guard let data = try? JSONEncoder.linkDiagnostics.encode(state) else {
      return
    }
    store.write(data)
  }

  private func snapshot(of state: PersistedState) -> SignalASILinkDiagnosticSnapshot {
    let counts = Dictionary(uniqueKeysWithValues: SignalASILinkDiagnosticKind.allCases.map { kind in
      (kind, state.counts[kind.rawValue] ?? 0)
    })
    return SignalASILinkDiagnosticSnapshot(
      totalEvents: state.totalEvents,
      counts: counts,
      recentEvents: Array(state.recentEvents.reversed())
    )
  }

  private struct PersistedState: Codable {
    var protocolName: String
    var totalEvents: Int64
    var counts: [String: Int64]
    var recentEvents: [SignalASILinkDiagnosticEvent]

    init(
      protocolName: String = SignalASILinkDiagnosticLedger.protocolVersion,
      totalEvents: Int64 = 0,
      counts: [String: Int64] = [:],
      recentEvents: [SignalASILinkDiagnosticEvent] = []
    ) {
      self.protocolName = protocolName
      self.totalEvents = totalEvents
      self.counts = counts
      self.recentEvents = recentEvents
    }

    enum CodingKeys: String, CodingKey {
      case protocolName = "protocol"
      case totalEvents = "total_events"
      case counts
      case recentEvents = "recent_events"
    }

    init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      protocolName = try container.decodeIfPresent(String.self, forKey: .protocolName) ??
        SignalASILinkDiagnosticLedger.protocolVersion
      totalEvents = try container.decodeIfPresent(Int64.self, forKey: .totalEvents) ?? 0
      counts = try container.decodeIfPresent([String: Int64].self, forKey: .counts) ?? [:]
      recentEvents = try container.decodeIfPresent(
        [SignalASILinkDiagnosticEvent].self,
        forKey: .recentEvents
      ) ?? []
    }
  }
}

enum SignalASILinkTransportDiagnostics {
  private static let lock = NSLock()
  private static var ledger: SignalASILinkDiagnosticLedger?

  static func runtimeLedger() -> SignalASILinkDiagnosticLedger {
    lock.lock()
    defer { lock.unlock() }
    if let ledger {
      return ledger
    }
    let next = SignalASILinkDiagnosticLedger(store: UserDefaultsSignalASILinkDiagnosticStore())
    ledger = next
    return next
  }

  @discardableResult
  static func record(
    kind: SignalASILinkDiagnosticKind,
    endpointIdentity: String = "",
    messageIdentity: String = "",
    detailCode: String = ""
  ) -> SignalASILinkDiagnosticSnapshot {
    runtimeLedger().record(
      kind: kind,
      endpointIdentity: endpointIdentity,
      messageIdentity: messageIdentity,
      detailCode: detailCode
    )
  }

  static func snapshot() -> SignalASILinkDiagnosticSnapshot {
    runtimeLedger().snapshot()
  }

  static func clear() {
    runtimeLedger().clear()
  }

  static func classifyDecryptionFailure(_ error: Error) -> SignalASILinkDiagnosticKind {
    let className = String(describing: type(of: error)).lowercased(with: Locale(identifier: "en_US_POSIX"))
    let message = "\(String(describing: error)) \((error as NSError).localizedDescription)"
      .lowercased(with: Locale(identifier: "en_US_POSIX"))
    if message.contains("old counter") || message.contains("oldcounter") {
      return .oldCounter
    }
    if className.contains("duplicatemessage") || message.contains("duplicate message") {
      return .duplicateMessage
    }
    return .decryptFailure
  }

  static func classifyFragmentFailure(_ error: Error) -> SignalASILinkDiagnosticKind {
    let message = "\(String(describing: error)) \((error as NSError).localizedDescription)"
      .lowercased(with: Locale(identifier: "en_US_POSIX"))
    return message.contains("duplicate") ? .chunkDuplicate : .fragmentRejected
  }
}

private extension JSONEncoder {
  static var linkDiagnostics: JSONEncoder {
    JSONEncoder()
  }
}

private extension JSONDecoder {
  static var linkDiagnostics: JSONDecoder {
    JSONDecoder()
  }
}
